crates/store/re_log_encoding/src/codec/file/decoder.rs

```diff
@@ -0,0 +1,107 @@
+        let mut buf = [0; 4];
+        data.read_exact(&mut buf)?;
+
+        match u32::from_le_bytes(buf) {
+            1 => Ok(Self::SetStoreInfo),
+            2 => Ok(Self::ArrowMsg),
+            3 => Ok(Self::BlueprintActivationCommand),
+            255 => Ok(Self::End),
+            _ => Err(DecodeError::Codec(CodecError::UnknownMessageHeader)),
+        }
+    }
+}
+
+impl MessageHeader {
+    pub(crate) fn decode(data: &mut impl std::io::Read) -> Result<Self, DecodeError> {
+        let kind = MessageKind::decode(data)?;
+        let mut buf = [0; 4];
+        data.read_exact(&mut buf)?;
+        let len = u32::from_le_bytes(buf);
+
+        Ok(Self { kind, len })
+    }
+}
+
+pub(crate) fn decode(
+    data: &mut impl std::io::Read,
+    compression: Compression,
+) -> Result<(u64, Option<LogMsg>), DecodeError> {
+    use re_protos::external::prost::Message;
+    use re_protos::log_msg::v0::{ArrowMsg, BlueprintActivationCommand, Encoding, SetStoreInfo};
+
+    let mut read_bytes = 0u64;
+    let header = MessageHeader::decode(data)?;
+    read_bytes += std::mem::size_of::<MessageHeader>() as u64 + header.len as u64;
+
+    let mut buf = vec![0; header.len as usize];
+    data.read_exact(&mut buf[..])?;
+
+    let msg = match header.kind {
+        MessageKind::SetStoreInfo => {
+            let set_store_info = SetStoreInfo::decode(&buf[..])?;
+            Some(LogMsg::SetStoreInfo(set_store_info.try_into()?))
+        }
+        MessageKind::ArrowMsg => {
+            let arrow_msg = ArrowMsg::decode(&buf[..])?;
+            if arrow_msg.encoding() != Encoding::ArrowIpc {
+                return Err(DecodeError::Codec(CodecError::UnsupportedEncoding));
+            }
+
+            let (schema, chunk) = decode_arrow(&arrow_msg.payload, compression)?;
+
+            let store_id: re_log_types::StoreId = arrow_msg
+                .store_id
+                .ok_or_else(|| {
+                    TypeConversionError::missing_field("rerun.log_msg.v0.ArrowMsg", "store_id")
+                })?
+                .into();
+
+            let chunk = re_chunk::Chunk::from_transport(&re_chunk::TransportChunk {
+                schema,
+                data: chunk,
+            })?;
+
+            Some(LogMsg::ArrowMsg(store_id, chunk.to_arrow_msg()?))
+        }
+        MessageKind::BlueprintActivationCommand => {
+            let blueprint_activation_command = BlueprintActivationCommand::decode(&buf[..])?;
+            Some(LogMsg::BlueprintActivationCommand(
+                blueprint_activation_command.try_into()?,
+            ))
+        }
+        MessageKind::End => None,
+    };
+
+    Ok((read_bytes, msg))
+}
+
+fn decode_arrow(
+    data: &[u8],
+    compression: crate::Compression,
+) -> Result<
+    (
+        arrow2::datatypes::Schema,
+        arrow2::chunk::Chunk<Box<dyn re_chunk::Arrow2Array>>,
+    ),
+    DecodeError,
+> {
+    let mut uncompressed = Vec::new();
+    let data = match compression {
```

Comment:
I have bad memory and don't remember if we decided anything on this one, but did we explore using arrow ipc compression?

Reply:
We have not experimented with this. I'd like to leave that out for now, created an issue:
- https://github.com/rerun-io/rerun/issues/8412